packages/@expo/cli/src/start/server/metro/inspector-proxy/proxy.ts

```diff
@@ -0,0 +1,155 @@
+   */
+  private setServerAddress(server: HttpServer | HttpsServer) {
+    const addressInfo = server.address();
+
+    if (typeof addressInfo === 'string') {
+      throw new Error(`Inspector proxy could not resolve the server address, got "${addressInfo}"`);
+    } else if (addressInfo === null) {
+      throw new Error(`Inspector proxy could not resolve the server address, got "null"`);
+    }
+
+    const { address, port, family } = addressInfo;
+
+    if (family === 'IPv6') {
+      this.metroProxy._serverAddressWithPort = `[${address ?? '::1'}]:${port}`;
+    } else {
+      this.metroProxy._serverAddressWithPort = `${address ?? 'localhost'}:${port}`;
+    }
+  }
+
+  /** @see https://chromedevtools.github.io/devtools-protocol/#endpoints */
+  public processRequest(req: IncomingMessage, res: ServerResponse, next: (error?: Error) => any) {
+    this.metroProxy.processRequest(req, res, next);
+  }
+
+  public createWebSocketListeners(server: HttpServer | HttpsServer): Record<string, WSServer> {
+    this.setServerAddress(server);
+
+    return {
+      [WS_DEVICE_URL]: this.createDeviceWebSocketServer(),
+      [WS_DEBUGGER_URL]: this.createDebuggerWebSocketServer(),
+    };
+  }
+
+  private createDeviceWebSocketServer() {
+    const wss = new WS.Server({
+      noServer: true,
+      perMessageDeflate: false,
+    });
+
+    // See: https://github.com/facebook/metro/blob/eeb211fdcfdcb9e7f8a51721bd0f48bc7d0d211f/packages/metro-inspector-proxy/src/InspectorProxy.js#L157
+    wss.on('connection', (socket, request) => {
+      try {
+        const deviceId = this.metroProxy._deviceCounter++;
+        const { deviceName, appName } = getNewDeviceInfo(request.url);
+
+        this.devices.set(
+          deviceId,
+          new this.DeviceClass(deviceId, deviceName, appName, socket, this.metroProxy._projectRoot)
+        );
+
+        debug('New device connected: device=%s, app=%s', deviceName, appName);
+
+        socket.on('close', () => {
+          this.devices.delete(deviceId);
+          debug('Device disconnected: device=%s, app=%s', deviceName, appName);
+        });
+      } catch (error: unknown) {
+        const message = error instanceof Error && error.toString();
+        console.error('Could not establish a connection to device:', error);
+        socket.close(WS_GENERIC_ERROR_STATUS, message || 'Unknown error');
+      }
+    });
+
+    return wss;
+  }
+
+  private createDebuggerWebSocketServer() {
+    const wss = new WS.Server({
+      noServer: true,
+      perMessageDeflate: false,
+    });
+
+    // See: https://github.com/facebook/metro/blob/eeb211fdcfdcb9e7f8a51721bd0f48bc7d0d211f/packages/metro-inspector-proxy/src/InspectorProxy.js#L193
+    wss.on('connection', (socket, request) => {
+      try {
+        const { deviceId, pageId } = getExistingDeviceInfo(request.url);
+        if (!deviceId || !pageId) {
+          throw new Error(`Missing "device" and/or "page" IDs in query parameters`);
```

Comment:
Add a TODO to come back to these and change them to CommandError or related. If these happen often, then we should be handling them more gracefully.

Reply:
```suggestion
          // TODO(cedric): change these errors to proper error types
          throw new Error(`Missing "device" and/or "page" IDs in query parameters`);
```